### DISCUSSION TITLE
add innerHTML input in the design panel

### DISCUSSION THREAD

**Discussion by surreal6** [uxbox/uxbox#602] [Ideas]
I would suggest to add an input for the innerHTML attribute in the design panel.  It will improve the experience of customice component instances.

Currently I need to lock some layers and groups to be able to double click on a text element inside a complex group, in order to customice it's content...

I have encouraged myself to try to add it to the code. A bit of help would be great, so:

- how do you manage the creation of the sidepanel?

- can you show me the portion of code in wich this occurs? 

thanks in advance and keep the good work!

**Comment by niwinz** [comment]
hello @surreal6 

First of all, welcome and thanks for writing us.

I'm not clearly undestand you. Correct me if I'm wrong: you want to be able edit raw svg of the component? That is immposible because internally we store all shapes as data structures (for make posble, colaborative edition, advanced transformations and so on...); so there are no innerHTML attribute at all.

About text edition on complex groups: we are already working on improvements for edition of text on complex groups; Leaving internal representation to the free edition by the user is not our goal right now, sorry.

**Reply by surreal6** [reply]
Hi @niwinz

you are right, mi question is not clear enought :P 
1.- about the propossal:
with the innerHTML attribute, sorry, (i'm mixing the naming between html and svg or even d3 for calling the same)
i mean the content, the string inside the tags of a text element in the final svg. I understand you have an internal model where you operate, and another component/factory/whatever to build the interface (i saw about react).

I understand the content of a text element is a property the same range it is color or round corner radius.

2.- about the derivated thoughts on the propossal :P
Have to tell i'm not a traditional developer but a creative one, and i learn a lot by hacking and tinkering.
What i am asking for is a guidance about the model and the general structure of the app, about the design patterns you follow or some help to read around the code to understand...

i'm not planning to add the element to the aside to solve something i'm pretty sure you'll fix in the future, with an absolutely better solution, but to modify the code to learn about it.

hope you understand my point :)

**Reply by hirunatan** [reply]
If I understand you, no: we don't store the svg code as you can see it in the "code mode" view, we generate it dynamically each time it's shown. So you can't customize it without touching the internal model.

If you want to see how is this model, you could look at the specs file https://github.com/penpot/penpot/blob/develop/common/app/common/pages/spec.cljc

And the code that generates the svg text from the shapes is here
https://github.com/penpot/penpot/tree/develop/frontend/src/app/main/ui/shapes

It's not very simple, I'm afraid. Basically, we store the model as some kind of "parsed" svg, i.e. one clojure object for each SVG node, with an id, a list of the attributes and the parent/child relationships. We have pending to document it in a more legible way and adding it to the Technical Guide.

**Reply by surreal6** [reply]
My explanation was not very accurate (it's kind of dificult to explain), but with this I already have where to start digging, so thank you for the links. The first one is not working, i think it should be this:

https://github.com/penpot/penpot/blob/develop/common/src/app/common/pages/spec.cljc

Also, I have found the technical guide very useful to get started reading the code:

https://help.penpot.app/technical-guide

**Reply by hirunatan** [reply]
Ok. I really hope that soon we can work in improving the technical guide, adding more sections to help to understand the application.
